### PR TITLE
Update flutter_tools to use package:file throughout

### PR DESCRIPTION
**Pull request by tvolkert**
This removes direct file access from within flutter_tools
in favor of using `package:file` via a `FileSystem` that's
accessed via the `ApplicationContext`.

This lays the groundwork for us to be able to easily swap
out the underlying file system when running Flutter tools,
which will be used to provide a record/replay file system,
analogous to what we have for process invocations.

### CODE REVIEW

**Comment by tvolkert** [comment]
FYI, this will require updating package:file to 1.0.1 since I added some functionality to it to support this change.  Until I update it, Travis will fail, but those failures can be ignored - I'll make sure it's passing before I submit.

File: packages/flutter_tools/bin/fuchsia_builder.dart [hunk 1]
```diff
@@ -3,13 +3,14 @@
 // found in the LICENSE file.
 
 import 'dart:async';
-import 'dart:io';
+import 'dart:io' as io;
```
Comment (Hixie) [inline packages/flutter_tools/bin/fuchsia_builder.dart]: consider importing it as `import 'dart:io' as io hide File, FileSystemEntity;` ?

Or maybe have a base/io.dart file that imports dart:io, and the file system and process packages, and re-exports the supported APIs?

The idea being to stop people accidentally depending on the APIs we don't want to use directly any more.

File: packages/flutter_tools/lib/src/base/file_system.dart [hunk 2]
```diff
@@ -2,45 +2,51 @@
...
 void restoreFileSystem() {
-  fs = new LocalFileSystem();
+  //context.setVariable(FileSystem, new LocalFileSystem());
   exit = _defaultExitFunction;
 }
 
 /// Uses in-memory replacments for `dart:io` functionality. Useful in tests.
 void useInMemoryFileSystem({ String cwd: '/', ExitFunction exitFunction }) {
-  fs = new MemoryFileSystem();
+  /*
```
Comment (abarth) [inline packages/flutter_tools/lib/src/base/file_system.dart]: Should we remove this code rather than comment it out?

Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/file_system.dart]: Oops - meant to do that before sending out the PR.  Done now.

File: packages/flutter_tools/bin/fuchsia_builder.dart [hunk 3]
```diff
@@ -37,6 +38,7 @@ Future<Null> main(List<String> args) async {
   executableContext.setVariable(Logger, new StdoutLogger());
   executableContext.runInZone(() {
     // Initialize the context with some defaults.
+    context.putIfAbsent(FileSystem, () => new LocalFileSystem());
```
Comment (Hixie) [inline packages/flutter_tools/bin/fuchsia_builder.dart]: I really would prefer dropping the use of zones here in favour of explicitly passing in a "context" variable of some sort, but that's a story for another day.

File: packages/flutter_tools/lib/src/base/file_system.dart [hunk 4]
```diff
@@ -2,45 +2,51 @@
...
+  io.exit(exitCode);
 };
 
 /// Exits the process.
 ExitFunction exit = _defaultExitFunction;
 
 /// Restores [fs] to the default local disk-based implementation.
 void restoreFileSystem() {
-  fs = new LocalFileSystem();
+  //context.setVariable(FileSystem, new LocalFileSystem());
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/base/file_system.dart]: ?

File: packages/flutter_tools/lib/src/base/file_system.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/base/file_system.dart]: ?

File: packages/flutter_tools/lib/src/base/file_system.dart (hunk 4 above)
Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/file_system.dart]: Already fixed in most recent commit

File: packages/flutter_tools/lib/src/base/file_system.dart (hunk 2 above)
Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/file_system.dart]: Already fixed in most recent commit

File: packages/flutter_tools/test/drive_test.dart [hunk 5]
```diff
@@ -52,7 +59,7 @@ void main() {
 
     tearDown(() {
       command = null;
-      restoreFileSystem();
+      restoreFileSystem(); // TODO: rename
```
Comment (Hixie) [inline packages/flutter_tools/test/drive_test.dart]: rename

Comment (Hixie) [inline packages/flutter_tools/test/drive_test.dart]: (or use correct TODO syntax)

Comment (tvolkert) [inline packages/flutter_tools/test/drive_test.dart]: Already done

**Comment by Hixie** [comment]
We use Platform.pathSeparator in a few places. Is that going to cause any trouble?

**Comment by Hixie** [comment]
RSLGTM

**Comment by tvolkert** [comment]
> We use Platform.pathSeparator in a few places. Is that going to cause any trouble?

Theoretically, it could.  I'll add a `FileSystem.pathSeparator` property and wire it up in a follow-on change (it'll only bite us for Windows users running tests, so it's not critical right now).

File: packages/flutter_tools/bin/fuchsia_builder.dart (hunk 1 above)
Comment (tvolkert) [inline packages/flutter_tools/bin/fuchsia_builder.dart]: ack.  Will do this in a separate change.

**Pull request merged**